components/scream/src/physics/p3/p3_functions_upwind_impl.hpp

```diff
@@ -0,0 +1,140 @@
+#ifndef P3_FUNCTIONS_UPWIND_IMPL_HPP
+#define P3_FUNCTIONS_UPWIND_IMPL_HPP
+
+#include "p3_functions.hpp"
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 upwind functions. Clients should NOT #include
+ * this file, #include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+template <Int kdir, int nfield>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::calc_first_order_upwind_step (
+  const ko::Unmanaged<view_1d<const Spack> >& rho,
+  const ko::Unmanaged<view_1d<const Spack> >& inv_rho,
+  const ko::Unmanaged<view_1d<const Spack> >& inv_dzq,
+  const MemberType& team,
+  const Int& nk, const Int& k_bot, const Int& k_top, const Scalar& dt_sub,
+  const view_1d_ptr_array<Spack, nfield>& flux,
+  const view_1d_ptr_array<Spack, nfield>& V,
+  const view_1d_ptr_array<Spack, nfield>& r)
+{
+  Int
+    kmin = ( kdir == 1 ? k_bot : k_top)             / Spack::n,
+    // Add 1 to make [kmin, kmax). But then the extra term (Spack::n -
+    // 1) to determine pack index cancels the +1.
+    kmax = ((kdir == 1 ? k_top : k_bot) + Spack::n) / Spack::n;
+  const Int k_top_pack = k_top / Spack::n;
+
+  // calculate fluxes
+  Kokkos::parallel_for(
+    Kokkos::TeamThreadRange(team, kmax - kmin), [&] (Int k_) {
```

Comment:
Again, I think it's worth to specify this is ok on GPU. E.g., something like "On GPU, these updates will be executed by all threads in the team vector length. That's fine, they are all simply going to write the same number. No race condition."